src/native/managed/cdacreader/src/Helpers/Metadata.cs

```diff
@@ -0,0 +1,340 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Diagnostics.DataContractReader.Contracts;
+using static System.Net.Mime.MediaTypeNames;
+
+namespace Microsoft.Diagnostics.DataContractReader.Helpers;
+
+internal enum MetadataTable
+{
+    Unused = -1,
+    Module = 0x0,
+    TypeRef = 0x01,
+    TypeDef = 0x02,
+    FieldPtr = 0x03,
+    Field = 0x04,
+    MethodPtr = 0x05,
+    MethodDef = 0x06,
+    ParamPtr = 0x07,
+    Param = 0x08,
+    InterfaceImpl = 0x09,
+    MemberRef = 0x0a,
+    Constant = 0x0b,
+    CustomAttribute = 0x0c,
+    FieldMarshal = 0x0d,
+    DeclSecurity = 0x0e,
+    ClassLayout = 0x0f,
+    FieldLayout = 0x10,
+    StandAloneSig = 0x11,
+    EventMap = 0x12,
+    EventPtr = 0x13,
+    Event = 0x14,
+    PropertyMap = 0x15,
+    PropertyPtr = 0x16,
+    Property = 0x17,
+    MethodSemantics = 0x18,
+    MethodImpl = 0x19,
+    ModuleRef = 0x1a,
+    TypeSpec = 0x1b,
+    ImplMap = 0x1c,
+    FieldRva = 0x1d,
+    ENCLog = 0x1e,
+    ENCMap = 0x1f,
+    Assembly = 0x20,
+    AssemblyProcessor = 0x21,
+    AssemblyOS = 0x22,
+    AssemblyRef = 0x23,
+    AssemblyRefProcessor = 0x24,
+    AssemblyRefOS = 0x25,
+    File = 0x26,
+    ExportedType = 0x27,
+    ManifestResource = 0x28,
+    NestedClass = 0x29,
+    GenericParam = 0x2a,
+    MethodSpec = 0x2b,
+    GenericParamConstraint = 0x2c,
+    Count = 0x2d
+}
+
+internal class EcmaMetadata
+{
+    public EcmaMetadata(EcmaMetadataSchema schema,
+                        ReadOnlyMemory<byte>[] tables,
+                        ReadOnlyMemory<byte> stringHeap,
+                        ReadOnlyMemory<byte> userStringHeap,
+                        ReadOnlyMemory<byte> blobHeap,
+                        ReadOnlyMemory<byte> guidHeap)
+    {
+        Schema = schema;
+        _tables = tables;
+        StringHeap = stringHeap;
+        UserStringHeap = userStringHeap;
+        BlobHeap = blobHeap;
+        GuidHeap = guidHeap;
+    }
+
+    public EcmaMetadataSchema Schema { get; init; }
+
+    private ReadOnlyMemory<byte>[] _tables;
+    public ReadOnlySpan<ReadOnlyMemory<byte>> Tables => _tables;
```

Comment:
Could we use `ReadOnlySequence<byte>` here? Basically, instead of having each element of the span represent the memory of a table, we could represent the underlying memory of the image as a `ReadOnlySequence<byte>` that stitches together the various components of the image.

You could use DNMD's implementation of the "save to memory" function to provide the remaining byte arrays to stitch together the separate tables and heaps into a valid ECMA-335 image represented as a `ReadOnlySequence<byte>`.

If we were to do this, we could (performance-permitting), refactor System.Reflection.Metadata to support this use case and remove the actual metadata reader implementation from here (and only have the "stitch into a sequence that represents metadata" code).

Reply:
> Could we use ReadOnlySequence<byte> here? Basically, instead of having each element of the span represent the memory of a table, we could represent the underlying memory of the image as a ReadOnlySequence<byte> that stitches together the various components of the image.

My gut reaction here would be introducing this abstraction is a bit of magic I'm not sure is a win. Do we have an example I can see that shows a benefit?

Reply:
I think that's not a bad idea, or at least, it doesn't really look like one. I don't really want to pursue it yet though.

1. This implementation should allow us to fill out the remaining metadata needs of the cDAC without all that much difficulty (other than the IMetadataImport export requirement, which System.Reflection.Metadata doesn't help with at all, in fact, I'm not sure the IMetadataImport api surface can be implemented on top of System.Reflection.Metadata).
2. I want to explore the possibility in .NET 10 of changing how we produce triage dump metadata when it becomes CDAC driven. Currently we produce an parallel data stream that has precomputed strings, but an alternative approach that might be more scalable to our real needs could be to have triage dump metadata be fully real metadata but only the exact bytes that the reader needs (not entire tables). Having a custom metadata reader makes adding that sort of hook a fair bit easier.

Reply:
System.Reflection.Metadata is the managed low-level metadata reader library. I do not think we want to be growing a different one. We should refactor System.Reflection.Metadata as needed.

Reply:
@jkotas, I don't see that as a practical near term effort. To make it able to handle the cases that come up here, would require replacing all of the memory handling of that library with something different, which I can't see happening as part of this effort until we've stabilized the capabilities of what this logic will require the metadata system to be able to process. In general, I view what I've built as a short term solution, that will either last until we replace it with dnmd, or we do a major refactor of S.R.M to make it able to operate on things that aren't a pointer to a contiguous memory blob.